src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceMessageBus.cs

```diff
@@ -2,24 +2,74 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
+using System.Linq;
 using System.Threading.Tasks;
 using Microsoft.AspNet.SignalR;
+using Microsoft.AspNet.SignalR.Infrastructure;
 using Microsoft.AspNet.SignalR.Messaging;
+using Microsoft.Azure.SignalR.Protocol;
 
 namespace Microsoft.Azure.SignalR.AspNet
 {
-    /// <summary>
-    /// TODO: This class is responsible for transform SignalR Messages to ServiceMessages and send them to the service runtime
-    /// </summary>
     internal class ServiceMessageBus : MessageBus
     {
+        private readonly IMessageParser _parser;
+        private readonly IServiceConnectionManager _serviceConnectionManager;
+        private readonly IAckHandler _ackHandler;
+
         public ServiceMessageBus(IDependencyResolver resolver) : base(resolver)
         {
+            _serviceConnectionManager = resolver.Resolve<IServiceConnectionManager>() ?? throw new ArgumentNullException(nameof(IServiceConnectionManager));
+            _parser = resolver.Resolve<IMessageParser>() ?? throw new ArgumentNullException(nameof(IMessageParser));
+            _ackHandler = resolver.Resolve<IAckHandler>() ?? throw new ArgumentNullException(nameof(IAckHandler));
         }
 
         public override Task Publish(Message message)
         {
-            throw new NotImplementedException();
+            var messages = _parser.GetMessages(message).ToList();
+            if (messages.Count == 0)
+            {
+                return Task.CompletedTask;
+            }
+
+            if (messages.Count == 1)
+            {
+                return ProcessMessage(messages[0]);
+            }
+
+            return Task.WhenAll(messages.Select(m => ProcessMessage(m)));
+        }
+
+        private Task ProcessMessage(AppMessage message)
+        {
+            if (message is HubMessage hubMessage)
+            {
+                return WriteMessage(_serviceConnectionManager.WithHub(hubMessage.HubName), message);
+            }
+
+            return WriteMessage(_serviceConnectionManager, message);
+        }
+
+        private async Task WriteMessage(IServiceConnectionContainer connection, AppMessage appMessage)
+        {
+            var message = appMessage.Message;
+            switch (message)
+            {
+                // For group related messages, make sure messages are written to the same partition
+                case JoinGroupMessage joinGroupMessage:
+                    await connection.WriteAsync(joinGroupMessage.GroupName, joinGroupMessage);
+                    _ackHandler.TriggerAck(appMessage.RawMessage.CommandId);
```

Comment:
This needs to always fire in case of a group command. Can WriteAsync ever throw?

Reply:
It is possible, wrap into a `finally` block